### PR TITLE
SLVS-1446 Add BindingDto model & serialization

### PR DESCRIPTION
**Pull request by georgii-borovinskikh-sonarsource**
[SLVS-1446](https://sonarsource.atlassian.net/browse/SLVS-1446)

[SLVS-1446]: https://sonarsource.atlassian.net/browse/SLVS-1446?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/ConnectedMode/Persistence/BindingDto.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+using Newtonsoft.Json;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.Persistence;
+
+[JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
+internal class BindingDto
```
Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode/Persistence/BindingDto.cs:29]: I would rather call it "BindingJsonModel" or "BindingJsonDto" to differentiate it between other dtos (used, for example, to communicate with SlCore) and one used for json

File: src/ConnectedMode.UnitTests/Persistence/BindingDtoConverterTests.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            ProjectKey = "project123",
+            Profiles = new Dictionary<Language, ApplicableQualityProfile>(),
+            ProjectName = "ignored",
+            Organization = new SonarQubeOrganization("ignored", "ignored"),
+            ServerUri = new Uri("http://ignored"),
+            ServerConnectionId = "ignored", // only used for connection extraction, delegates to the connection object during the conversion
+        };
+        var connection = new ServerConnection.SonarCloud("myorg");
+
+        var boundServerProject = new BindingDtoConverter().ConvertFromDto(bindingDto, connection, localBindingKey);
```
Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode.UnitTests/Persistence/BindingDtoConverterTests.cs:59]: Please extract all object creation of BindingDtoConverter into a variable in TestInitialize.

File: src/ConnectedMode/Persistence/BindingDtoConverter.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            ServerConnectionId = binding.ServerConnection.Id,
+            Profiles = binding.Profiles,
+            // for compatibility reasons:
+            ServerUri = binding.ServerConnection.ServerUri,
+            Organization = binding.ServerConnection is ServerConnection.SonarCloud sonarCloudConnection
+                ? new SonarQubeOrganization(sonarCloudConnection.OrganizationKey, null)
+                : null
+        };
+
+    public BoundSonarQubeProject ConvertFromDtoToLegacy(BindingDto bindingDto, ICredentials credentials) =>
```
Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode/Persistence/BindingDtoConverter.cs:57]: Why are we expecting here that the BindingDto is null? But not in the other convert methods?

Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode/Persistence/BindingDtoConverter.cs:57]: Also there is no test case when BindingDto is null

File: src/ConnectedMode.UnitTests/Persistence/BindingDtoSerializationTests.cs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    {
+        var convertedFromDtoToLegacy = bindingDtoConverter.ConvertFromDtoToLegacy(JsonConvert.DeserializeObject<BindingDto>(JsonConvert.SerializeObject(bindingDto)), null);
+        
+        convertedFromDtoToLegacy.Should().BeEquivalentTo(boundSonarQubeProject);
+    }
+    
+    [TestMethod]
+    public void CurrentSonarCloud_ToDto_ToJson_ToDto_ToLegacy_IsCorrect()
+    {
+        var legacyBinding = bindingDtoConverter.ConvertFromDtoToLegacy(JsonConvert.DeserializeObject<BindingDto>(JsonConvert.SerializeObject(bindingDtoConverter.ConvertToDto(boundSonarCloudServerProject))), null);
```
Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode.UnitTests/Persistence/BindingDtoSerializationTests.cs:172]: Please extract all the train wrecks into multiple variables, to make the test more readable.

File: src/ConnectedMode.UnitTests/Persistence/BindingDtoSerializationTests.cs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        
+        legacyBinding.Should().BeEquivalentTo(boundSonarQubeProject, options => options.Excluding(x => x.ProjectName).Excluding(x => x.ServerUri).Excluding(x => x.Organization.Name));
+        legacyBinding.ServerUri.Should().BeEquivalentTo(boundSonarCloudServerProject.ServerConnection.ServerUri);
+        legacyBinding.ProjectName.Should().BeNull();
+        legacyBinding.Organization.Name.Should().BeNull();
```
Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode.UnitTests/Persistence/BindingDtoSerializationTests.cs:177]: Missing the check for OrganizationKey

File: src/ConnectedMode/Persistence/BindingDto.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public string ServerConnectionId { get; set; }
+    public Uri ServerUri { get; set; } // left here for compatibility reasons
```
Comment (gabriela-trutan-sonarsource) [inline src/ConnectedMode/Persistence/BindingDto.cs:32]: Maybe you can group the three compatibility properties at the bottom and add a better comment explaining what does "compatibility reason" means. Example of a better comment "The information for these properties are now stored in a separate file for connections, but they are needed here for backward compatibility with previous binding formats"

**Review by gabriela-trutan-sonarsource — CHANGES_REQUESTED** [review]

File: src/ConnectedMode/Persistence/BindingDto.cs (hunk 1 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/Persistence/BindingDto.cs:29]: Yeah I wanted to do that, but then thought maybe json is too specific. I can rename it, sure

File: src/ConnectedMode/Persistence/BindingDtoConverter.cs (hunk 3 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode/Persistence/BindingDtoConverter.cs:57]: Could be an error that was introduced when I was extracting this method

File: src/ConnectedMode.UnitTests/Persistence/BindingDtoSerializationTests.cs (hunk 5 above)
Comment (georgii-borovinskikh-sonarsource) [reply src/ConnectedMode.UnitTests/Persistence/BindingDtoSerializationTests.cs:177]: Separate checks are only performed for excluded properties. Here, only Organization.Name is excluded. Organization.Key is a part of the equivalence assertion

File: src/ConnectedMode/Persistence/BindingDto.cs (hunk 6 above)
Comment (gabriela-trutan-sonarsource) [reply src/ConnectedMode/Persistence/BindingDto.cs:32]: I see. But then maybe just improve the comment?

**Review by gabriela-trutan-sonarsource — APPROVED** [review]
